### PR TITLE
First version

### PR DESCRIPTION
**Pull request by AFrancani**
[sc-63042]

### CODE REVIEW

File: python-runnables/export-to-stages/runnable.json [hunk 1]
```diff
@@ -0,0 +1,102 @@
+/* This file is the descriptor for the python runnable export-to-stages */
+{
+    "meta": {
+        // label: name of the runnable as displayed, should be short
```
Comment (apichery) [inline python-runnables/export-to-stages/runnable.json]: All these scaffolding comments should be removed imo.

File: python-runnables/export-to-stages/runnable.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+# This file is the actual code for the Python runnable export-to-stages
+import logging
+from dataiku import SQLExecutor2, Dataset, set_default_project_key
+from dataiku.runnables import Runnable
+
+
+class MyRunnable(Runnable):
```
Comment (apichery) [inline python-runnables/export-to-stages/runnable.py]: Do we have to call the call `MyRunnable` or could we use a more meaningful name?

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,51 @@
+pipeline {
```
Comment (HenriChabert) [inline Jenkinsfile]: If you don't have test, you can remove the Jenkinsfile as it is only used for Jenkins integration

File: plugin.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+    "id": "snowflake-stages-export",
+    "version": "0.0.1",
+    "meta": {
```
Comment (HenriChabert) [inline plugin.json:4]: I could be a good idea to assign a category to the plugin. 
`"category": "Cloud connection"

File: python-lib/snowflakestagesexport/__init__.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# When creating plugins, it is a good practice to put the specific logic in libraries and keep plugin components (recipes, etc) short. 
+# You can add functionalities to this package and/or create new packages under "python-lib"
```
Comment (HenriChabert) [inline python-lib/snowflakestagesexport/__init__.py]: Can be removed

File: resource/compute_dynamic_params.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        except Exception as e:
+            logging.exception('Error while fetching Snowflake file formats for dataset `%s`', dataset_name, exc_info=e)
+            choices = [failed_connection("file formats")]
+        return {"choices": choices}
+
+
+def get_snowflake_datasets():
+    project_key = default_project_key()
+    project = api_client().get_project(project_key)
+    return filter(lambda dataset: dataset.type == 'Snowflake', project.list_datasets())
```
Comment (HenriChabert) [inline resource/compute_dynamic_params.py]: In Python, we prefer the structure as the structure you use returns a `filter object`
```suggestion
    return [ds for ds in project.list_datasets() if ds.type == 'Snowflake']
```

File: resource/compute_dynamic_params.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        connection = dataset['params']['connection']
+        if connection not in datasets_per_connection:
+            datasets_per_connection[connection] = []
+        datasets_per_connection[connection] += [dataset.name]
+
+    # Only if there are multiple Snowflake connections, we need to display the different inputs grouped by connections
+    multiple_connections = len(datasets_per_connection) > 1
+
+    # indent only if we have multiple Snowflake connections
+    indent = (lambda choice: {'value': choice['value'], 'label': f"  {choice['label']}"}) if multiple_connections else (lambda choice: choice)
```
Comment (HenriChabert) [inline resource/compute_dynamic_params.py]: Not a big fan of this kind of structure hard to read. I would rather right that:
```suggestion
    def do_indentation(choice):
        return {'value': choice['value'], 'label': f"  {choice['label']}"}
    indent = do_indentation if multiple_connections else (lambda choice: choice)
```

Comment (HenriChabert) [inline resource/compute_dynamic_params.py]: I would also give a value to the number of spaces you set when indenting to be able to change it later.
```suggestion
    indent = (lambda choice: {'value': choice['value'], 'label': f"{" "*3}{choice['label']}"}) if multiple_connections else (lambda choice: choice)
```

File: python-runnables/export-to-stages/runnable.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        for param in mandatory_params:
+            if param['id'] not in self.config or not self.config.get(param['id']):
+                logging.error('The mandatory param `%s` is missing or invalid to export dataset `%s.%s` to Snowflake stage',
+                              param['name'], self.project_key, dataset_name)
+                raise ValueError(f"The parameter '{param['name']}' is invalid")
+
+        fully_qualified_stage_name = self.config.get('stage')
+
+        output_path = (self.config.get('path') or self.project_key).strip(' ').strip('/')
+        destination = f"{output_path}/{dataset_name}" if output_path else dataset_name
```
Comment (HenriChabert) [inline python-runnables/export-to-stages/runnable.py]: The `os.path` can do all that for you:
```suggestion
        output_path = (self.config.get('path') or self.project_key)
        destination = os.path.join(output_path, dataset_name) if output_path else dataset_name
```

File: python-runnables/export-to-stages/runnable.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+        file_format_param = self.config.get('file_format') or 'default'
```
Comment (HenriChabert) [inline python-runnables/export-to-stages/runnable.py:67]: Use the get structure:
```suggestion
        file_format_param = self.config.get('file_format', 'default')
```

File: python-runnables/export-to-stages/runnable.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        overwrite = 'OVERWRITE = TRUE' if self.config.get("overwrite") else ''
+
+        sql_copy_query = f"COPY INTO @{fully_qualified_stage_name}/{destination} FROM {resolve_table_name(dataset_connection_info)} {file_format} {overwrite}"
+
+        logging.info("Exporting dataset `%s.%s` with the copy command: `%s`", self.project_key, dataset_name, sql_copy_query)
+
+        executor = SQLExecutor2(dataset=dataset_name)
+        executor.query_to_df(sql_copy_query)
+
+        logging.info("Successfully exported dataset `%s.%s` in Snowflake stage `%s` to `%s`",
```
Comment (HenriChabert) [inline python-runnables/export-to-stages/runnable.py]: Use f-string:
```suggestion
        logging.info(f"Successfully exported dataset `{self.project_key}.{dataset_name}` in Snowflake stage `{fully_qualified_stage_name}` to `{destination}`",
```

**Review by HenriChabert — APPROVED** [review]
Everything looks very good to me, great work :)

Some suggestions though, mainly minors

File: plugin.json (hunk 4 above)
Comment (AFrancani) [reply plugin.json:4]: What is the point of the category exactly? Couldn't find where it is used.

File: resource/compute_dynamic_params.py (hunk 7 above)
Comment (AFrancani) [reply resource/compute_dynamic_params.py]: This is a temporary manual indentation which will be removed when dynamic select inputs are released.

File: python-runnables/export-to-stages/runnable.py (hunk 9 above)
Comment (AFrancani) [reply python-runnables/export-to-stages/runnable.py:67]: The behavior is a bit different:
```python
>>> config = dict()
>>> config['a'] = 1
>>> config['a']
1
>>> config['b'] = None
>>> config['b']
>>> config['c'] = ''
>>> config.get('a') or 'default'
1
>>> config.get('b') or 'default'
'default'
>>> config.get('c') or 'default'
'default'
>>> config.get('d') or 'default'
'default'
>>> config.get('a', 'default')
1
>>> config.get('b', 'default')
>>> config.get('c', 'default')
''
>>> config.get('d', 'default')
'default'
```

**Pull request merged**